File: api/app/models/ExamComponent.py
from datetime import datetime

# request body data validation
from pydantic import BaseModel

class ExamComponentBase(BaseModel):
    exam_id: int
    exam_component_type_id: int
    mark: int=0
    comment: str=None
    page_start: int
    page_end: int
    sequence: int

class ExamComponentNewIn(ExamComponentBase):
    pass

class ExamComponentIn(ExamComponentBase):
    id: int=None

class ExamComponentOut(ExamComponentBase):
    id: int
    course_id: int
    created: datetime
    modified: datetime

File: ui/src/models/ExamComponent.js
import {Model, Collection} from 'vue-mc'
import {getAPIURL, convertDatabaseDate} from './Helpers'

export class ExamComponent extends Model {
  defaults() {
    return {
      id: null,
      course_id: null,
      exam_id: null,
      exam_component_type_id: null,
      exam_component_type: '',
      sequence: null,
      page_start: 0,
      page_end: 0,
      mark: 0,
      comment: '',
      created: '',
      modified: ''
    }
  }

  mutations() {
    return {
      id:   (id) => Number(id) || null,
      created: (created) => convertDatabaseDate(created),
      modified: (modified) => convertDatabaseDate(modified)
    }
  }

  validation() {
    return {}
  }

  routes() {
    return {
      fetch: getAPIURL('/courses/{course_id}/exams/{exam_id}/components/{id}'),
      create: getAPIURL('/courses/{course_id}/exams/{exam_id}/components'),
      save: getAPIURL('/courses/{course_id}/exams/{exam_id}/components/{id}'),
      delete: getAPIURL('/courses/{course_id}/exams/{exam_id}/components/{id}')
    }
  }
}

export class ExamComponentList extends Collection {
  defaults() {
    return {
      course_id: null,
      exam_id: null
    }
  }

  model() {
    return ExamComponent
  }

  routes() {
    return {
      fetch: getAPIURL('/courses/{course_id}/exams/{exam_id}/components'),
    }
  }
}
